drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -551,22 +542,88 @@ static int32_t adpd410x_get_data_packet(struct adpd410x_dev *dev,
 		}
 	} while(i < no_slots);
 
-	free(data_byte_buff);
-	free(data_buff);
 	free(slot_bytecount_tab);
-
 	return SUCCESS;
 
-error_data_byte:
-	free(data_byte_buff);
-error_data_buff:
-	free(data_buff);
 error_slot:
 	free(slot_bytecount_tab);
 
 	return ret;
 }
 
+/**
+ * @brief Reads a certain number of bytes from the fifo and stores in data
+ *        Used to read a large amount of data from the fifo efficiently (using
+ *        as few register reads as possible.)
+ * @param dev - Device handler.
+ * @param data - Pointer to the data container.
+ * @param num_samples - number of samples to read
+ * @param datawidth - number of bytes per sample
+ * @return SUCCESS in case of success, FAILURE or an error code otherwise.
+ */
+int32_t adpd410x_read_fifo(struct adpd410x_dev *dev, uint32_t *data,
+			   uint16_t num_samples,
+			   uint8_t datawidth)
+{
+	int32_t ret = SUCCESS;
+	uint8_t *data_byte_buff, i, j;
+	uint16_t next_packet_size, bytes_read = 0,
+				   total_bytes = num_samples * datawidth;
+	if (datawidth > 4 || total_bytes > ADPD410X_FIFO_DEPTH)
+		return FAILURE;
+
+	data_byte_buff = (uint8_t *) calloc(total_bytes, sizeof (*data_byte_buff));
+	if (!data_byte_buff) {
+		return -ENOMEM;
+	}
+
+	while (bytes_read < total_bytes) {
+		// Can read a maximum of 255 bytes at once
+		next_packet_size = (total_bytes - bytes_read > 255) ? 255 : total_bytes -
+				   bytes_read;
+		ret = adpd410x_reg_read_bytes(dev, ADPD410X_REG_FIFO_DATA,
+					      data_byte_buff + bytes_read,
+					      next_packet_size);
+		if(ret != SUCCESS)
+			goto fifo_free_return;
+
+		bytes_read += next_packet_size;
+	}
+
+	if (data == NULL)
```

Comment:
Would it make more sense to check this at the beginning, before doing any reads?